feat(oxlint): add `--experimental-nested-config` option

camchenry opened the pull request:
- part of https://github.com/oxc-project/oxc/issues/7408

Since the nested configuration work is too big to fit in one PR or even a stack of PRs, I suggest that we make it temporarily an opt-in CLI flag. Once we feel confident in the implementation replacing the existing code, we can remove the flag and make it true by default.

camchenry commented:
* **#9153** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9153?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9152** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9152?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9152?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9152).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Sysix commented on tasks/website/src/linter/snapshots/cli_terminal.snap:
```diff
@@ -81,5 +81,7 @@ Available positional items:
 
 Available options:
         --rules               list all the rules that are currently registered
+        --experimental-nested-config  Enables automatic loading of nested configuration files
```
NIT: its not aligned with the other options :/

Boshen replied:
This is bpaf's problem :-/

pacak replied:
> This is bpaf's problem :-/

Intentional design decision :) Always aligning to the longest option name means we can get in a situation where you have a word or two on a line (if trying to keep overall help width small-ish) or making everything super wide - not very readable either. 

Current behavior means making an attempt to align everything, but allowing long options to be ugly.

Boshen replied:
This is actually fine once we remove the `experimental` prefix.

camchenry replied:
I don't think it should be a long-term problem: I think this option should become a default eventually. If anything, maybe we add a `--no-nested-config` option to disable it, but we'll definitely drop the experimental part.

Sysix replied:
> If anything, maybe we add a --no-nested-config option to disable it, 
I could not find a similar eslint flag. But there is a `root` key in the configuration file

camchenry replied:
I think the equivalent is `--no-eslintrc` or `--no-config-lookup`:

https://eslint.org/docs/latest/use/command-line-interface#--no-eslintrc

camchenry committed:
feat(oxlint): add `--experimental-nested-config` option (#9152)

- part of https://github.com/oxc-project/oxc/issues/7408

Since the nested configuration work is too big to fit in one PR or even a stack of PRs, I suggest that we make it temporarily an opt-in CLI flag. Once we feel confident in the implementation replacing the existing code, we can remove the flag and make it true by default.
